Support executors in Slurm and HTCondor batch system adapters

stefan-k opened the pull request:
This PR allows users to specify and executor (ShellExecutor or SSHExecutor, default is the first) for the Slurm and HTCondor batch system adapters. Documentation and tests were also updated accordingly. Due to the needed `mock_executor_run_command` decorator in the tests, it was necessary move some data from the test class into global variables. 

We don't have an HTCondor setup, therefore I was only able to verify the Slurm part.

giffels reviewed (changes requested):
Thanks a lot for your contribution. Could you address the comments made inline the code, please? 

Feel free discuss potential solutions for the comments about `async_run_command` in this pull request first. 

I currently see two possibilities: 

- Either we move the workaround for exit code 255 and `os.waitpid` called in a different thread to the `ShellExecutor` or
- we need to keep `async_run_command` and use the already existing `executor` argument.

stefan-k commented on tardis/adapters/batchsystems/htcondor.py:
```diff
@@ -17,7 +17,7 @@
 
 async def htcondor_status_updater(
-    options: AttributeDict, attributes: AttributeDict
+    options: AttributeDict, attributes: AttributeDict, executor=ShellExecutor
```
I wanted a fallback executor in case none is provided (which should not happen, so it is pointless) but apparently I forgot the brackets it instantiate the class. `executor: Executor` makes the most sense.

stefan-k commented on tardis/adapters/batchsystems/htcondor.py:
```diff
@@ -47,9 +47,9 @@ async def htcondor_status_updater(
     try:
         logger.debug(f"HTCondor status update is running. Command: {cmd}")
-        condor_status = await async_run_command(cmd)
+        condor_status = await executor.run_command(cmd)
```
I followed the implementation in the HTCondor site adapter: https://github.com/MatterMiners/tardis/blob/b3ea6fa926048c7812632a43663f0a5b005ee443/tardis/adapters/sites/htcondor.py#L32
Is there a reason why `async_run_command` is not used there?

giffels replied:
No, there is no reason. Probably the best way to solve it by removing `async_run_command` and implementing the workaround in the `ShellExecutor`. Most like nobody run into this issue using the HTCondor site adapter yet, because either modern Python versions (>3.7) are used or the `SSHExecutor` is used.

stefan-k replied:
If this workaround only affects `ShellExecutor`, then I also think this would be the best idea. If it is fine for you, I could do this as part of this PR or as part of a new one. Might take a while though because I'll probably be fighting the tests again.

giffels replied:
Yes, it only affects affects the `ShellExecutor`. Would be awesome if you could do that in this pull request.

stefan-k commented on tardis/adapters/batchsystems/slurm.py:
```diff
@@ -64,7 +64,7 @@ async def slurm_status_updater(
     except CommandExecutionFailure as ex:
         logging.warning(f"SLURM's sinfo could not be executed! {str(ex)}")
-        raise
+        raise ex
```
Thanks, that's just a remnant of a desperate debugging session.

stefan-k commented on tests/adapters_t/batchsystems_t/test_htcondor.py:
```diff
@@ -90,53 +90,60 @@ def test_disintegrate_machine(self):
             run_async(self.htcondor_adapter.disintegrate_machine, drone_uuid="test")
         )
 
+    @mock_executor_run_command(stdout=CONDOR_RETURN)
```
`drain_machine` executes a status update before calling `condor_drain`: https://github.com/MatterMiners/tardis/blob/b3ea6fa926048c7812632a43663f0a5b005ee443/tardis/adapters/batchsystems/htcondor.py#L122
When removing this decorator, the test will fail:
```
>           condor_status = await executor.run_command(cmd)
E           TypeError: object MagicMock can't be used in 'await' expression
```
As you may have noticed, I'm not a mocking expert, so it is not unlikely that something else is also wrong in the testing setup.

stefan-k commented on tests/adapters_t/batchsystems_t/test_htcondor.py:
```diff
@@ -90,53 +90,60 @@ def test_disintegrate_machine(self):
-        self.mock_async_run_command.side_effect = None
+        self.mock_executor.return_value.run_command.side_effect = None
 
+    @mock_executor_run_command(stdout=CONDOR_RETURN)
```
Interestingly, this decorator can be removed without issues....

stefan-k commented on tests/adapters_t/batchsystems_t/test_slurm.py:
```diff
@@ -80,53 +81,59 @@ def test_disintegrate_machine(self):
             run_async(self.slurm_adapter.disintegrate_machine, drone_uuid="test")
         )
 
+    @mock_executor_run_command(stdout=SINFO_RETURN)
```
Same issue as in the HTCondor tests, it fails when it is removed because the status update is called.

giffels commented:
Something went wrong with assigning the reviewers yesterday. Originally @maxfischer2781  and @giffels  got assigned.

stefan-k commented:
As discussed in the review, I've removed `async_run_command` and included its logic into `ShellExecutor`. Essentially I'm treating `exit_code` 0 and 255 as succeeded and raise `CommandExecutionFailure` otherwise. The `ShellExecutor` tests were adapted accordingly. The Slurm and HTCondor batch system adapters were also changed back to directly calling `run_command` of `self._executor`.

mschnepf reviewed (approved):
Looks good to me. :+1:

maxfischer2781 commented on tests/adapters_t/batchsystems_t/test_slurm.py:
```diff
@@ -9,29 +11,43 @@
+from unittest.mock import patch
 from unittest import TestCase
 
+SINFO_RETURN = "\n".join(
+    [
+        "mixed      2/2/0/4   6000    24000   VM-1   host-10-18-1-1",
+        "mixed      3/1/0/4   15853   22011   VM-2   host-10-18-1-2",
+        "mixed      1/3/0/4   18268   22011   VM-3   host-10-18-1-4",
+        "mixed      3/1/0/4   17803   22011   VM-4   host-10-18-1-7",
+        "draining   0/4/0/4   17803   22011   draining_m   draining_m",
+        "idle       0/4/0/4   17803   22011   idle_m   idle_m",
+        "drained    0/4/0/4   17803   22011   drained_m   drained_m",
+        "powerup    0/4/0/4   17803   22011   pwr_up_m   pwr_up_m",
+    ]
+)
+
```
Now that this is at the top-level, you should be able to conveniently use a multi-line literal:
```suggestion
SINFO_RETURN = """\
mixed      2/2/0/4   6000    24000   VM-1   host-10-18-1-1
mixed      3/1/0/4   15853   22011   VM-2   host-10-18-1-2
mixed      1/3/0/4   18268   22011   VM-3   host-10-18-1-4
mixed      3/1/0/4   17803   22011   VM-4   host-10-18-1-7
draining   0/4/0/4   17803   22011   draining_m   draining_m
idle       0/4/0/4   17803   22011   idle_m   idle_m
drained    0/4/0/4   17803   22011   drained_m   drained_m
powerup    0/4/0/4   17803   22011   pwr_up_m   pwr_up_m
""""

```

stefan-k replied:
Thanks for the suggestion, it is much more convenient this way.

maxfischer2781 reviewed (changes requested):
Got a minor cleanup for the unittests. Looks good to me otherwise.

maxfischer2781 reviewed (approved):
Good to go! 👍

giffels commented on tardis/adapters/batchsystems/slurm.py:
```diff
@@ -131,7 +133,7 @@ async def drain_machine(self, drone_uuid: str) -> None:
         cmd = f"scontrol update NodeName={machine} State=DRAIN Reason='COBalD/TARDIS'"
 
-        return await async_run_command(cmd)
+        return await self._executor.run_command(cmd)
```
According to the docstring this function should return `None`. It was returning `stdout` before, the return value of `async_run_command`. I would suggest to you remove the `return` in front of `await self._executor.run_command` to make it consistent to the documentation.

giffels reviewed (changes requested):
Thanks a lot for improving the code. I have only two minor things to fix before accepting. Please, see comments in code.

giffels reviewed (commented):
Thanks a lot. Now we are ready to merge!
